x/epochs/keeper/abci.go

```diff
@@ -0,0 +1,85 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cosmossdk.io/x/epochs/types"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+)
+
+// BeginBlocker of epochs module.
+func (k Keeper) BeginBlocker(ctx context.Context) error {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
+
+	logger := k.Logger()
+	err := k.EpochInfo.Walk(
+		ctx,
+		nil,
+		func(key string, epochInfo types.EpochInfo) (stop bool, err error) {
+			// If blocktime < initial epoch start time, return
+			if k.environment.HeaderService.GetHeaderInfo(ctx).Time.Before(epochInfo.StartTime) {
```

Comment:
nit: store k.environment.HeaderService.GetHeaderInfo(ctx) in a variable to avoid calling this many times